File: 0332-reconstruct-itinerary/0332-reconstruct-itinerary.js
/**
 * @param {string[][]} tickets
 * @return {string[]}
 */
var findItinerary = function(tickets) {
    const graph = {};

    for (const [from, to] of tickets) {
        if(!graph[from]) graph[from] = [];

        graph[from].push(to);
    }

    for (let node in graph) {
        graph[node].sort();
    }

    const dfs = function(curr) {
        const neighbors = graph[curr] || [];

        while(neighbors.length > 0) {
            const node = neighbors.shift();
            dfs(node);
        }
        
        path.push(curr);
    }

    const path = [];
    dfs('JFK');

    return path.reverse();
};

File: 0332-reconstruct-itinerary/0332-reconstruct-itinerary.py
class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        graph = {}
        path = []

        for start, end in tickets:
            if start not in graph:
                graph[start] = []
            graph[start].append(end)

        for node in graph:
            graph[node].sort()

        def dfs(curr):
            neighbors = graph.get(curr, [])
            while neighbors:
                node = neighbors.pop(0)
                dfs(node)
            path.append(curr)

        dfs("JFK")

        return path[::-1]